lib/AutoPodArgsPass.cpp

```diff
@@ -104,12 +119,24 @@ void AutoPodArgsPass::runOnFunction(Function &F) {
   // 1. Clustered pod args.
   // 2. No global push constants.
   // 3. Args must fit in push constant size limit.
+  // 4. No arrays.
+  // 5. If 16-bit types are used, 16-bit push constants are supported.
+  // 6. If 8-bit types are used, 8-bit push constants are supported.
   const auto pod_struct_ty = StructType::get(M.getContext(), pod_types);
+  const bool contains_array = ContainsArrayType(pod_struct_ty);
+  const bool support_16bit_pc = !ContainsSizedType(pod_struct_ty, 16) ||
+                                clspv::Option::Supports16BitStorageClass(
+                                    clspv::Option::StorageClass::kPushConstant);
+  const bool support_8bit_pc = !ContainsSizedType(pod_struct_ty, 8) ||
+                               clspv::Option::Supports8BitStorageClass(
+                                   clspv::Option::StorageClass::kPushConstant);
   const bool satisfies_push_constant =
-      !(!clspv::Option::ClusterPodKernelArgs() ||
-        clspv::UsesGlobalPushConstants(M) ||
+      clspv::Option::ClusterPodKernelArgs() && support_16bit_pc &&
```

Comment:
This is pretty hard to unpack. 
Recommend breaking down that last clause.  Can be a future CL.